src/main/scala/mesosphere/marathon/core/storage/impl/BasePersistenceStore.scala

```diff
@@ -0,0 +1,141 @@
+import mesosphere.util.LockManager
+
+import scala.async.Async.{ async, await }
+import scala.concurrent.{ ExecutionContext, Future }
+
+/**
+  * Persistence Store that handles all marshalling and unmarshalling, allowing
+  * subclasses to focus on the raw formatted data.
+  *
+  * @tparam K The persistence store's primary key type
+  * @tparam Serialized The serialized format for the persistence store.
+  */
+abstract class BasePersistenceStore[K, Category, Serialized](implicit
+  ctx: ExecutionContext,
+  mat: Materializer) extends PersistenceStore[K, Category, Serialized]
+    with TimedPersistenceStore[K, Category, Serialized] {
+
+  private[this] lazy val lockManager = LockManager.create()
+
+  protected def rawIds(id: Category): Source[K, NotUsed]
+
+  override def ids[Id, V]()(implicit ir: IdResolver[Id, V, Category, K]): Source[Id, NotUsed] = {
+    rawIds(ir.category).map(ir.fromStorageId)
+  }
+
+  protected def rawVersions(id: K): Source[OffsetDateTime, NotUsed]
+
+  final override def versions[Id, V](
+    id: Id)(implicit ir: IdResolver[Id, V, Category, K]): Source[OffsetDateTime, NotUsed] = {
+    rawVersions(ir.toStorageId(id, None))
+  }
+
+  protected def rawDelete(k: K, version: OffsetDateTime): Future[Done]
+
+  override def delete[Id, V](
+    k: Id,
+    version: OffsetDateTime)(implicit ir: IdResolver[Id, V, Category, K]): Future[Done] = {
+    val storageId = ir.toStorageId(k, Some(version))
+    lockManager.executeSequentially(k.toString) {
+      rawDelete(ir.toStorageId(k, Some(version)), version)
+    }
+  }
+
+  protected def rawDeleteAll(k: K): Future[Done]
+
+  final override def deleteAll[Id, V](k: Id)(implicit ir: IdResolver[Id, V, Category, K]): Future[Done] = {
+    val storageId = ir.toStorageId(k, None)
+    lockManager.executeSequentially(k.toString) {
+      rawDeleteAll(ir.toStorageId(k, None))
+    }
+  }
+
+  protected[storage] def rawGet(k: K): Future[Option[Serialized]]
+
+  override def get[Id, V](id: Id)(implicit
+    ir: IdResolver[Id, V, Category, K],
+    um: Unmarshaller[Serialized, V]): Future[Option[V]] = async {
+    val storageId = ir.toStorageId(id, None)
+    await(rawGet(storageId)) match {
+      case Some(v) =>
+        Some(await(Unmarshal(v).to[V]))
+      case None =>
+        None
+    }
+  }
+
+  override def get[Id, V](id: Id, version: OffsetDateTime)(implicit
+    ir: IdResolver[Id, V, Category, K],
+    um: Unmarshaller[Serialized, V]): Future[Option[V]] = async {
+    val storageId = ir.toStorageId(id, Some(version))
+    await(rawGet(storageId)) match {
+      case Some(v) =>
+        Some(await(Unmarshal(v).to[V]))
+      case None =>
+        None
+    }
+  }
+
+  protected def deleteOld(k: K, maxVersions: Int): Future[Done] = async {
+    val versions = await(rawVersions(k).toMat(Sink.seq)(Keep.right).run()).sortBy(_.toEpochSecond)
+    // we always store the current version (twice), once as a versioned node and once as the current one.
```

Comment:
This was kind of a weird oddity but it actually avoids a lot of extra gets and makes store() a bit easier - since its expected to keep the existing one if there are any.

versions is supposed to list the current version as well as the historical (I didn't originally do it this way) and that's consistent with the current MarathonStore.